db/column_family.cc

```diff
@@ -1176,6 +1176,105 @@ Status ColumnFamilyData::RangesOverlapWithMemtables(
   return status;
 }
 
+Status ColumnFamilyData::GetMemtablesUserKeyRange(PinnableSlice* smallest,
+                                                  PinnableSlice* largest,
+                                                  bool* found) {
+  assert(smallest && largest && found);
+  Status s;
+  auto* ucmp = user_comparator();
+  Arena arena;
+  ReadOptions read_opts;
+  read_opts.total_order_seek = true;
+  MergeIteratorBuilder merge_iter_builder(&internal_comparator_, &arena);
+  merge_iter_builder.AddIterator(mem_->NewIterator(read_opts, &arena));
+  imm_.current()->AddIterators(read_opts, &merge_iter_builder);
+  ScopedArenaIterator memtable_iter(merge_iter_builder.Finish());
+  ParsedInternalKey ikey;
+  memtable_iter->SeekToFirst();
+  if (memtable_iter->Valid()) {
+    s = ParseInternalKey(memtable_iter->key(), &ikey, false);
+    if (s.ok()) {
+      if (!(*found) || ucmp->Compare(ikey.user_key, *smallest) < 0) {
+        smallest->PinSelf(ikey.user_key);
+      }
+      memtable_iter->SeekToLast();
+      assert(memtable_iter->Valid());
+      s = ParseInternalKey(memtable_iter->key(), &ikey, false);
+      if (s.ok()) {
+        if (!(*found) || ucmp->Compare(*largest, ikey.user_key) < 0) {
+          largest->PinSelf(ikey.user_key);
+        }
+      }
+      *found = true;
+    }
+  }
+
+  if (s.ok()) {
+    autovector<MemTable*> memtables{mem_};
+    imm_.ExportMemtables(&memtables);
+    for (auto* mem : memtables) {
+      auto* iter =
+          mem->NewRangeTombstoneIterator(read_opts, kMaxSequenceNumber);
+      if (iter != nullptr) {
+        iter->SeekToFirst();
+        if (iter->Valid()) {
+          Slice start = iter->start_key();
```

Comment:
Why not use `iter->key()`, you have already sought to first

Reply:
And `start_key` is not a user_key, seems you should use `parsed_start_key`?

Reply:
> Why not use iter->key(), you have already sought to first

It's `FragmentedRangeTombstoneIterator`, the `key()` returns the start_key of current range tombstone, so we must use `end_key()` to get the upper bound, using `iter->key()` for the lower bound is okayish but not very consistent.

Reply:
> And start_key is not a user_key, seems you should use parsed_start_key?

Actually `start_key` for `FragmentedRangeTombstoneIterator` is a user key, it directly uses it to construct a `ParsedInternalKey`:
```c++
  ParsedInternalKey parsed_start_key() const {
    return ParsedInternalKey(pos_->start_key, kMaxSequenceNumber,
                             kTypeRangeDeletion);
  }
```